README.md
# Sensu Go Perl Runtime Assets
![release](https://github.com/nixwiz/sensu-go-fatigue-check-filter/workflows/release/badge.svg)

This project provides [Sensu Go Assets][sensu-assets] containing portable Perl
runtimes (for various platforms), based on [Sensu Ruby Runtime][sensu-ruby-runtime]
which itself was based on the excellent [ruby-install project
by postmodern][ruby-install]. In practice, this Perl runtime asset should allow
Perl-based scripts to be packaged as separate assets containing Perl scripts and
any corresponding module dependencies. In this way, a single shared Perl runtime
may be delivered to systems running the Sensu Go Agent via the Sensu's Asset framework
(i.e. avoiding solutions that would require a Perl runtime to be redundantly
packaged with every perl-based plugin).

[sensu-assets]: https://docs.sensu.io/sensu-go/latest/reference/assets/
[sensu-ruby-runtime]: https://github.com/sensu/sensu-ruby-runtime
[ruby-install]: https://github.com/postmodern/ruby-install

## Platform Coverage

Currently this repository only supports a subset of Linux distribution by making
use of Docker containers to build and test.  If you would like extend the coverage,
please take a look at the Github Action and test build scripts. I'm happy
to take pull requests that extending the platform coverage.  Here's the current
platform matrix that we are testing for as of the current release:

| Asset Platform                    | Tested Operating Systems Docker Images        |
|:----------------------------------|:----------------------------------------------|
|  alpine  (based on alpine:3.8)    | Alpine(3, 3.8, latest)                        |
|  centos7 (based on centos:7)      | CentOS(7), Oracle Linux(7)                    |
|  rocky8  (based on rockylinux:8)  | Rocky Linux(8), Oracle Linux(8)               |
|  amzn2   (based on amazonlinux:2) | Amazon Linux(2)                               |
|  debian9 (based on debian:9)      | Debian(8, 9, 10), Ubuntu(14.04, 16.04, 18.04) |

## Modules Included

The following modules (and their dependencies) are packaged as part of the runtime:
* AutoLoader
* DBI
* Data::Dumper
* DateTime
* Digest::MD5
* File::Basename
* File::Slurp
* Getopt::Long
* HTTP::Response
* IO::File
* IO::Socket::SSL
* JSON
* JSON::XS
* LWP
* LWP::Protocol::https
* LWP::UserAgent
* Module::Load
* Net::SMTP
* Net::SNMP<sup>1</sup>
* Test::Simple
* Text::CSV
* Time::Zone
* WWW::Mechanize
* XML::LibXML

1. The Net::SNMP module is not included for Alpine as I gave up on trying to get it to work in the Docker build environment.

## Using the assets (non-Bonsai)

### Dowloadable releases

Releases are currently available on [Github][https://github.com/nixwiz/sensu-perl-runtime/releases].

### Building the assets locally

1. Use a Docker container to build Perl, and generate a local_build Sensu Go Asset.

   ```
   $ docker build --build-arg "PERL_VERSION=5.34.0" -t sensu-perl-runtime:5.34.0-debian -f Dockerfile.debian .
   ```

2. Extract your new sensu-perl asset, and get the SHA-512 hash for your Sensu asset!

   ```
   $ mkdir dist
   $ docker run -v "$PWD/dist:/dist" sensu-perl-runtime:5.34.0-debian cp /assets/sensu-perl-runtime_local-build_perl-5.34.0_debian_linux_amd64.tar.gz /dist/
   $ shasum -a 512 dist/sensu-perl-*
   ```
### Using the assets

1. Put that asset somewhere that your Sensu agent can fetch it. Perhaps add it to the Bonsai asset index!

2. Create an asset resource in Sensu Go.

   First, create a configuration file called `sensu-perl-runtime-5.34.0-debian.json` with
   the following contents:

   ```
   {
     "type": "Asset",
     "api_version": "core/v2",
     "metadata": {
       "name": "sensu-ruby-runtime-5.34.0-debian",
       "namespace": "default",
       "labels": {},
       "annotations": {}
     },
     "spec": {
       "url": "http://your-asset-server-here/assets/sensu-perl_local-build_perl-5.34.0_debian_linux_amd64.tar.gz",
       "sha512": "4f926bf4328fbad2b9cac873d117f771914f4b837c9c85584c38ccf55a3ef3c2e8d154812246e5dda4a87450576b2c58ad9ab40c9e2edc31b288d066b195b21b",
       "filters": [
         "entity.system.os == 'linux'",
         "entity.system.arch == 'amd64'",
         "entity.system.platform == 'debian'"
       ]
     }
   }
   ```

   Then create the asset via:

   ```
   $ sensuctl create -f sensu-perl-runtime-5.34.0-debian.json
   ```

3. Create a second asset containing a Perl script.

   To run a simple test using the Perl runtime asset, create another asset
   called `helloworld-v0.1.tar.gz` with a simple perl script at
   `bin/helloworld.pl`; e.g.:

   ```perl
   #!/usr/bin/env perl

   use DateTime;

   $dt = DateTime->now;

   print "Hello world! The date and time is now " . $dt->stringify() . "\n";

   ```

   _NOTE: this is a simple "hello world" example, but it shows that we have
   support for basic perl modules!_

   Compress this file into a g-zipped tarball and register this asset with
   Sensu, and then you're all ready to run some tests!

4. Create a check resource in Sensu Go.

   First, create a configuration file called `helloworld.json` with
   the following contents:

   ```
   {
     "type": "CheckConfig",
     "api_version": "core/v2",
     "metadata": {
       "name": "helloworld",
       "namespace": "default",
       "labels": {},
       "annotations": {}
     },
     "spec": {
       "command": "helloworld.pl",
       "runtime_assets": ["sensu-perl-runtime-5.34.0-debian", "helloworld-v0.1"],
       "publish": true,
       "interval": 10,
       "subscriptions": ["docker"]
     }
   }
   ```

   Then create the asset via:

   ```
   $ sensuctl create -f helloworld.json
   ```

   At this point, the `sensu-backend` should begin publishing your check
   request. Any `sensu-agent` member of the "docker" subscription should
   receive the request, fetch the Perl runtime and helloworld assets,
   unpack them, and successfully execute the `helloworld.pl` command by
   resolving the Perl shebang (`#!/usr/bin/env perl`) to the Perl runtime
   on the Sensu agent `$PATH`.
   
## Additional modules

TBD on how to make use of modules not currently included.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [1.3.0] - 2022-03-21

### Changed
- Install net-snmp modules by default (except alpine)
- Fix README to reflect above and clean up

## [1.2.1] - 2022-03-18

### Changed
- Transitioned from Travis CI to Github Actions
- Replaced CentOS 8 with Rocky Linux 8 for building for RHEL 8 and derivatives
- Replace [] and [[ ]] with test across all because Debian 9 /bin/sh is braindead (had to use /usr/bin/test to avoid shell builtin)

### Added
- Amazon Linux 2 support

### Removed
- CentOS 6 due to EOL
- Redundant Alpine config

## [1.1.0] - 2022-03-17

### Changed
- Updated Perl version to 5.34.0

### Added
- NETSNMP_INSTALL option
- add File-Slurp and JSON-XS module

## [1.0.6] - 2020-01-16

### Changed
- Updated build testing to include Oracle Linux

## [1.0.5] - 2020-01-13

### Changed
- Fixed Perl build to use -Duserelocatableinc so it can be relocated

## [1.0.4] - 2020-01-09

### Changed
- Fixed Alpine build to be more in line with Ruby Runtime build

## [1.0.3] - 2020-01-08

### Changed
- More fixes for mismatched github and docker.io account names
- Fix checksum script that I messed up by removing a line I thought was unnecessary

## [1.0.2] - 2020-01-08

### Changed
- Fix problem when mismatched github and docker.io account names

## [1.0.1] - 2020-01-08

### Changed
- Some minor cosmetic changes

## [1.0.0] - 2020-01-07

### Added
- Initial release


scripts/test.sh
#!/bin/sh

install_tar() {
  # Please don't look at me, I'm hideous
  if type yum > /dev/null 2>&1; then
    yum -y install tar gzip
  elif type apt-get > /dev/null 2>&1; then
    apt-get install -y tar
  elif type apk > /dev/null 2>&1; then
    apk --no-cache add tar
  fi
}

echo "Test Script:"
echo "  Asset Platform:  ${platform}"
echo "  Target Platform: ${test_platform}"
echo "  Asset Tarball:   ${asset_filename}"
if [ -z "$asset_filename" ]; then
  echo "Asset is empty"
  exit 1
fi
mkdir -p /build
cd /build
# Some Docker images (I'm talking to you Oracle Linux) don't
# include tar, so go find it
if [ ! -x /bin/tar ] && [ ! -x /usr/bin/tar ]; then
  install_tar
fi
tar xzf /dist/$asset_filename
LD_LIBRARY_PATH="/build/lib:$LD_LIBRARY_PATH" /build/bin/perl /scripts/test_ssl_url.pl



scripts/test_ssl_url.pl
use LWP::UserAgent ();

print "Checking HTTPS response with verify disabled\n";

my $ua = LWP::UserAgent->new(timeout => 10);
$ua->env_proxy;
$ua->ssl_opts( verify_hostnames => 0 );

my $response = $ua->get("https://www.sensu.io/");

if ($response->is_success) {
    print "Success!\n";
}
else {
    die $response->status_line;
}



build_and_test_platform.sh
#!/bin/bash

ignore_errors=0
perl_version=5.34.0
asset_version=${TAG:-local-build}
asset_filename=sensu-perl-runtime_${asset_version}_perl-${perl_version}_${platform}_linux_amd64.tar.gz
asset_image=sensu-perl-runtime-${perl_version}-${platform}:${asset_version}

if [ "${asset_version}" = "local-build" ]; then
  echo "Local build"
  ignore_errors=1
fi

echo "Platform: ${platform}"
echo "Check for asset file: ${asset_filename}"
if [ -f "$PWD/dist/${asset_filename}" ]; then
  echo "File: "$PWD/dist/${asset_filename}" already exists!!!"
  [ $ignore_errors -eq 0 ] && exit 1  
else
  echo "Check for docker image: ${asset_image}"
  if [[ "$(docker images -q ${asset_image} 2> /dev/null)" == "" ]]; then
    echo "Docker image not found...we can build"
    echo "Building Docker Image: sensu-perl-runtime:${perl_version}-${platform}"
    docker build --build-arg "PERL_VERSION=$perl_version" --build-arg "ASSET_VERSION=$asset_version" --build-arg "MAKE_TEST_CMD=true" --build-arg "CPANM_TEST_FLAG=--notest" -t ${asset_image} -f Dockerfile.${platform} .
    echo "Making Asset: /assets/sensu-perl-runtime_${asset_version}_perl-${perl_version}_${platform}_linux_amd64.tar.gz"
    docker run --rm -v "$PWD/dist:/dist" ${asset_image} cp /assets/${asset_filename} /dist/
  else
    echo "Image already exists!!!"
    [ $ignore_errors -eq 0 ] && exit 1  
  fi
fi

test_arr=($test_platforms)
for test_platform in "${test_arr[@]}"; do
  echo "Test: ${test_platform}"
  docker run --rm -e platform -e test_platform=${test_platform} -e asset_filename=${asset_filename} -v "$PWD/scripts/:/scripts" -v "$PWD/dist:/dist" ${test_platform} /scripts/test.sh
  retval=$?
  if [ $retval -ne 0 ]; then
    echo "!!! Error testing ${asset_filename} on ${test_platform}"
    exit $retval
  fi
done

if [ -z "$TAG" ]; then exit 0; fi
if [ -z "$DOCKER_USER" ]; then exit 0; fi
if [ -z "$DOCKER_PASSWORD" ]; then exit 0; fi

# In the event that of mismatch between github and docker.io usernames
GITHUB_USER=$(echo $REPO_SLUG | cut -d/ -f1)
if [ "${GITHUB_USER}" = "${DOCKER_USER}" ]; then
  DOCKER_SLUG=${REPO_SLUG}
else
  GITHUB_REPO=$(echo $REPO_SLUG | cut -d/ -f2)
  DOCKER_SLUG=${DOCKER_USER}/${GITHUB_REPO}
fi

docker_asset=${DOCKER_SLUG}-${perl_version}-${platform}:${asset_version}

echo "Docker Hub Asset: ${docker_asset}"
echo "preparing to tag and push docker hub asset"

echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USER" --password-stdin

docker tag ${asset_image} ${docker_asset}
docker push ${docker_asset}

ver=${asset_version%+*}
prefix=${ver%-*}
prerel=${ver/#$prefix}
if [ -z "$prerel" ]; then 
  echo "tagging as latest asset"
  latest_asset=${DOCKER_SLUG}-${perl_version}-${platform}:latest
  docker tag ${asset_image} ${latest_asset}
  docker push ${latest_asset}
fi



build_platforms.sh
#!/bin/bash

mkdir -p dist
mkdir -p assets
mkdir -p scripts

# Debian platforms
platform="debian9" test_platforms="debian:8 debian:9 debian:10 ubuntu:14.04 ubuntu:16.04 ubuntu:18.04" ./build_and_test_platform.sh
retval=$?
if [[ retval -ne 0 ]]; then
  exit $retval
fi

# Alpine platform
platform="alpine3.8" test_platforms="alpine:latest alpine:3 alpine:3.8" ./build_and_test_platform.sh
retval=$?
if [[ retval -ne 0 ]]; then
  exit $retval
fi

# RHEL and derivative platforms
platform="centos7" test_platforms="centos:7 oraclelinux:7" ./build_and_test_platform.sh
retval=$?
if [[ retval -ne 0 ]]; then
  exit $retval
fi

platform="rocky8" test_platforms="rockylinux:8 centos:8 oraclelinux:8" ./build_and_test_platform.sh
retval=$?
if [[ retval -ne 0 ]]; then
  exit $retval
fi

# Amazon Linux
platform="amzn2" test_platforms="amazonlinux:2" ./build_and_test_platform.sh
retval=$?
if [[ retval -ne 0 ]]; then
  exit $retval
fi
